### DISCUSSION TITLE
Purpose of aSwapIndex for flash platform functions

### DISCUSSION THREAD

**Discussion by atar13** [openthread/openthread#9872] [Q&A] [answered]
Hello!

I am porting OpenThread to an [embedded OS](https://github.com/tock/tock) and am in the midst of porting flash functionality defined in this [`flash.h`](https://github.com/openthread/openthread/blob/f0b6fcea6ef77c9a54ab11767593f9a8798e3662/include/openthread/platform/flash.h).

I am struggling to figure out what the `aSwapIndex` parameter means for several of the flash functions. Here is the signature of one of them:
```cpp
/**
 * Reads @p aSize bytes into @p aData.
 *
 * @param[in]  aInstance   The OpenThread instance structure.
 * @param[in]  aSwapIndex  A value in [0, 1] that indicates the swap space.
 * @param[in]  aOffset     A byte offset within the swap space.
 * @param[out] aData       A pointer to the data buffer for reading.
 * @param[in]  aSize       Number of bytes to read.
 *
 */
void otPlatFlashRead(otInstance *aInstance, uint8_t aSwapIndex, uint32_t aOffset, void *aData, uint32_t aSize);
```

I have already tried looking around came across [this PR](https://github.com/openthread/openthread/pull/775). As I understand it, the swap is used as a buffer for writing settings. However, I was confused when I came across a [flash implementation](https://github.com/openthread/ot-nrf528xx/blob/887a0ea4b22d60204172d213f63c46e35130be41/src/src/flash.c) that just stored swap data in another region of flash. I don't understand why you would buffer to the same storage medium. Originally I thought it would make sense to store swap in RAM, however the PR I linked advises against that to avoid loss of data during power outage.

Could anyone explain what `aSwapIndex` is intended for and how I should implement its functionality?

Pinging @jwhui since they are the ones who committed this file (b3d05af6ffc9e743254a7d0f6aeaf91bfbf6544e).

Thank you!

**Comment by jwhui** [comment, marked as answer]
Flash storage typically performs erase operations on units of pages. To support atomic updates to settings in non-volatiles storage with flash, we switch between different swap spaces on flash by writing new data to a different page and then updating the pointer to the latest page. The swap index indicates which swap space to use.
